### PR TITLE
[Ready] FLIGHTPACK PATCH 2 - PROBABLY CONTAINS LARGE AMOUNTS OF SHITTY CODE

### PR DESCRIPTION
**Pull request by silicons**
:cl:
experimental: Flightsuits now have their own subsystem!
bugfix: Flightsuits properly account for power before calculating drifting
experimental: Flightpack users will automatically fly over anyone buckled without crashing.
experimental: Flightpack users automatically slip through mineral doors
experimental: Flightpack users will crash straight through grills at appropriate times 
experimental: Flightpack users automatically slip through unbolted airlocks
experimental: Flightpacks are faster in space, but their space momentum decay has been upped significantly to compensate
experimental: Flighthelmets now have a function to allow the wearer to zoom out to see further. Helps you not crash eh?

/:cl:

### CODE REVIEW

**Comment by GunHog** [comment]
Add a way (if there is not already) to repair the "damage" that occurs from crashing too much! Mine got to a point where  turning it on resulted in immediate stunning and shut-off.

**Comment by kevinz000** [comment]
@GunHog happens automatically per process tick

File: code/modules/clothing/spacesuits/flightsuit.dm [hunk 1]
```diff
@@ -799,6 +820,7 @@
 /obj/item/clothing/shoes/flightshoes/Destroy()
 	if(suit)
 		suit.shoes = null
+	..()
```
Comment (ChangelingRain) [inline code/modules/clothing/spacesuits/flightsuit.dm]: return ..()

File: code/modules/clothing/spacesuits/flightsuit.dm [hunk 2]
```diff
@@ -155,14 +160,24 @@
...
+	sleep(2)
+	//STOP_PROCESSING(SSfastprocess, src)
 	part_manip = null
 	part_scan = null
 	part_cap = null
 	part_laser = null
 	part_bin = null
 	..()
 
+/obj/item/device/flightpack/proc/bullshit_processing_proc()
```
Comment (RemieRichards) [inline code/modules/clothing/spacesuits/flightsuit.dm]: Not happening in 3 billion years.
Your own nonsense processing is faster than the crazily efficient load-balancing Master Controller? I think not.

**Review by RemieRichards — CHANGES_REQUESTED** [review]

**Comment by jughu** [comment]
Add thermals to the suit

File: code/modules/clothing/spacesuits/flightsuit.dm (hunk 2 above)
Comment (kevinz000) [inline code/modules/clothing/spacesuits/flightsuit.dm]: Fixed :)

**Comment by kevinz000** [comment]
ree

**Comment by GunHog** [comment]
Hope you tested this...

**Comment by kevinz000** [comment]
nevermind FUCK BUGS ;-;

**Comment by kevinz000** [comment]
@RemieRichards 's review has been addressed btw honk

**Comment by AnturK** [comment]
@kevinz000 This is ready for review ?

File: code/modules/clothing/spacesuits/flightsuit.dm [hunk 3]
```diff
@@ -492,6 +520,38 @@
 		losecontrol(move = FALSE)
 	crashing = FALSE
 
+/obj/item/device/flightpack/proc/door_hit(obj/structure/mineral_door/door)
```
Comment (AnturK) [inline code/modules/clothing/spacesuits/flightsuit.dm]: These things are done this way instead of some generic function overrides for a reason, anyway ?

**Review by forkingpath — CHANGES_REQUESTED** [review]
```
[10:31:17]Cannot execute null.visible message().
[10:31:17]proc name: losecontrol (/obj/item/device/flightpack/proc/losecontrol)
[10:31:17]  source file: flightsuit.dm,555
[10:31:17]  usr: null
[10:31:17]  src: the flight pack (/obj/item/device/flightpack/full)
[10:31:17]  src.loc: Angelina Holmes (/mob/living/carbon/human)
[10:31:17]  call stack:
[10:31:17]the flight pack (/obj/item/device/flightpack/full): losecontrol(1, 1)
[10:31:17]the flight pack (/obj/item/device/flightpack/full): disable flight(1)
[10:31:17]the flight pack (/obj/item/device/flightpack/full): check conditions()
[10:31:17]the flight pack (/obj/item/device/flightpack/full): process(2)
```

edit: this is VERY frequent by the way, almost as frequent as /turf/var/excited runtime

https://tgstation13.org/parsed-logs/basil/runtimes/runtime-2016-12-19.txt

File: code/modules/clothing/spacesuits/flightsuit.dm (hunk 3 above)
Comment (kevinz000) [inline code/modules/clothing/spacesuits/flightsuit.dm]: I don't think there's a generic function I can override...

**Comment by kevinz000** [comment]
@karlnp Pretty sure that's because one of the admins called some proc on her doesn't happen normally.

File: code/modules/clothing/spacesuits/flightsuit.dm (hunk 3 above)
Comment (AnturK) [inline code/modules/clothing/spacesuits/flightsuit.dm]: That's when you add new one. Cost of a function on atom/movable is negligible and makes this much easier to expand and change.

**Comment by jughu** [comment]
Just waiting for mekhi to make a var to edit flightspeed or no fail flight

File: code/modules/clothing/spacesuits/flightsuit.dm (hunk 3 above)
Comment (optimumtact) [inline code/modules/clothing/spacesuits/flightsuit.dm]: ehhhhhhhhhhh I'm not sold on that Anturk, I can't see many things needing to interact with doors like this

File: code/modules/clothing/spacesuits/flightsuit.dm [hunk 4]
```diff
@@ -492,6 +520,38 @@
...
+		pass += A.locked
+		pass += A.stat	//No power, no automatic open
+		pass += A.emagged
+		if(A.requiresID())
+			if((!A.allowed(wearer)) && !A.emergency)
+				pass += 1
+	else
+		return pass
+	if(!pass)
+		spawn()
```
Comment (optimumtact) [inline code/modules/clothing/spacesuits/flightsuit.dm]: why are you doing this in a spawn()?

File: code/modules/clothing/spacesuits/flightsuit.dm (hunk 3 above)
Comment (AnturK) [inline code/modules/clothing/spacesuits/flightsuit.dm]: True enough, this is niche use.

File: code/modules/clothing/spacesuits/flightsuit.dm (hunk 4 above)
Comment (kevinz000) [inline code/modules/clothing/spacesuits/flightsuit.dm]: Bugs out if I don't I assume it waits for it to open completely before letting you through which ends up in you teleporting back to the door after you fly through.

**Pull request merged**